packages/compose/src/hooks/use-dragging/index.js

```diff
@@ -0,0 +1,74 @@
+/**
+ * WordPress dependencies
+ */
+import {
+	useCallback,
+	useEffect,
+	useLayoutEffect,
+	useRef,
+	useState,
+} from '@wordpress/element';
+
+const useIsomorphicLayoutEffect =
+	typeof window !== 'undefined' ? useLayoutEffect : useEffect;
```

Comment:
@jorgefilipecosta any context for why we introduced this? It's unclear to me what this is supposed to do, especially since the React docs mention that neither `useLayoutEffect` nor `useEffect` can run until downloading JS in the browser.

Reply:
Hi @dmsnell, if we use useLayoutEffect on a component and that component is server-side rendered or rendered to a string using renderToString there will be a warning/error. So we use useIsomorphicLayoutEffect to try to make it possible to have our components used in projects relying on server-side render.

Reply:
thanks for the clarification @jorgefilipecosta 

two follow-ups:

 - why swap out `useEffect` since according to the React docs [it has the same issues on server render](https://reactjs.org/docs/hooks-reference.html#uselayouteffect) as `useLayoutEffect` does, namely that it won't run on the server because they have to wait until the browser has loaded and parsed the JS.

 - why hide the warning? isn't that there to alert developers that their components won't likely behave the way they think if rendered from the server? am I misunderstanding or are we simply hiding that warning from people while preserving the same risk?

Reply:
Well, I think in this case if we server-side render a component even if the effect does not run the component renders the expected HTML just without the events so hiding the warning is a good thing.
But, I remember I explored at the time some open source code for similar situations/event handling and the useIsomorphicLayoutEffect was common practice so I just followed it, if you think it is unnecessary in this case we can remove it.

Reply:
> if you think it is unnecessary in this case we can remove it.

Might do that if we can be reasonably sure it's not a depended-upon API.
The way I think through it I think I'd rather leave that warning to alert people of the potential issues, which I think are basically that if you perform any state mutation in those effects that you expect to appear on the first render then they won't be there on server renders, and that's true for both functions, so I am confused why we make the switch based on whether `window` exists, unless again it's just to hide the warning.

After a bit of searching I'm finding the whole topic a big ambiguous in the official docs and declarations. It appears like the only reason `useEffect` _doesn't_ throw a warning on server render is more happenstance than intent: it simply never runs on the server and so doesn't get a chance to complain.

> to make it possible to have our components used in projects relying on server-side render

what prevents someone from using this in server-side render context if it spits out a warning?

Reply:
> Might do that if we can be reasonably sure it's not a depended-upon API.

On this specific component, useIsomorphicLayoutEffect was not exposed and we can safely not use it. But it seems after this work a useIsomorphicLayoutEffect hook was added and was exposed. We can remove the use cases of the hook when it makes sense.